sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -178,11 +179,50 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
   }
 
   /**
-   * Add an INSERT INTO [TABLE]/INSERT OVERWRITE TABLE operation to the logical plan.
+   * Parameters used for writing query to a directory: (isLocal, CatalogStorageFormat, provider).
+   */
+  type InsertDirParams = (Boolean, CatalogStorageFormat, Option[String])
+
+  /**
+   * Add an
+   *   INSERT INTO [TABLE] or
+   *   INSERT OVERWRITE TABLE or
+   *   INSERT OVERWRITE [LOCAL] DIRECTORY
+   * operation to logical plan
    */
   private def withInsertInto(
       ctx: InsertIntoContext,
       query: LogicalPlan): LogicalPlan = withOrigin(ctx) {
+    assert(ctx.children.size == 1)
+
+    ctx.getChild(0) match {
+      case c if c. isInstanceOf[InsertIntoTableContext] =>
```

Comment:
`match` can do type detection.. this can be simplified as :

```
      case tableContext: InsertIntoTableContext => withInsertIntoTable(tableContext, query)
      case dirContext: InsertOverwriteDirectoryContext => withInsertOverwriteDirectory(dirContext, query)
```

Reply:
updated